update catalog.py

Devhub01 opened the pull request:
edit/delete of domains, authors, tags

Devhub01 committed:
update catalog.py

edit/delete of domains, authors, tags

Devhub01 committed:
edit page

This html page serves to edit the items : domains, authors, tags

Devhub01 committed:
Update db.py

Update tables: papers_domains, papers_authors, papers_keywords, after deletion.
Check domain existence in db before adding it (case insensible)

Commit:
update papers-in-domains-authors-keywords

kerzol commented on templates/catalog/papers-in-domain.html:
```diff
@@ -1,5 +1,33 @@
+
+
+
+<h4>Papers in domain <em>{{domainname}}</em></h4>
```
I propose to use the following code here. It will be more concise.
If you are ok with this modification, could you also change the related templates for tags and authors in the same manner?

```html
<h4>Papers in domain <em>{{domainname}}</em>
{% if user_authenticated() %}
 {% if entry.domainname | can_edit_domain %}
  [<a title="Delete the domain" onclick="return confirm('You really want to delete the domain ?')"
    href="/domain/delete/{{domainname}}">remove</a> /
  <a title="Modify the domain name  " href="/domain/edit/{{domainname}}">edit</a>]
 {% endif %}
{% endif %}
</h4>
```

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
-    keywords=query_db ("select * from keywords order by keyword")
-    authors=query_db ("select * from authors order by fullname")
-    users=query_db ("select * from users where valid = 1 or userid = 1 order by username")
+    domains = query_db("select * from domains order by domainname")
```
It is ok for this time. But I strongly advise you to only include in the pull request only the "real" changes. The syntax improvements should go to another pull request.

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+                          papers=with_description(papers))
+
+
+####################### My Modifications ##############################
```
The line is useful for this review. But it can disturb the future reader, because "my" may be related to anyone who participated in the project :)

Devhub01 replied:
I modified this to "modification made by my GitHub username", so it will help me keep track of what I did in order to write the PFE report, but you can delete this line when merging the branches.

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+                                    where domainid = ?',
+                            [request.form['name'], request.form['id']])
+
+                # TODO:
```
TODO what ?

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+                                        domainname=request.form['name']))
+    return render_template('catalog/edit.html',
+                           entry=domain,
+
```
Please, remove that empty line

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+                                    where keywordid = ?',
+                            [request.form['name'], request.form['id']])
 
+                # TODO:
```
same remark as before

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+                                    where authorid = ?',
+                            [request.form['name'], request.form['id']])
+
+                # TODO:
```
same remark as before

kerzol commented on papersite/db.py:
```diff
@@ -141,19 +154,23 @@ def get_insert_author(fullname):
         con.execute("INSERT OR IGNORE INTO domains(domainname)    \
-                     VALUES(?)", [domainname])
+                    SELECT (?) WHERE not exists                     \
+                    (select (?) from domains where domainname = (?) COLLATE NOCASE)",(domainname,domainname,domainname))
```
Bien vu! Superb!

Devhub01 replied:
Thank you :D

kerzol commented on templates/catalog/edit.html:
```diff
@@ -0,0 +1,134 @@
+
+{% block scripts %}
+
+<script>
```
I'm not sure that the autocomplete list is useful here.
Imagine you have two domains named respectively D1 and D2. 
As the user edits the first domain, he can write D2 instead of D1.
And this will generate the error:
![u2](https://user-images.githubusercontent.com/748084/99193538-f92e0180-2779-11eb-8afb-753e12e942f8.png)

It happens because domain (tag, authors) names are UNIQUE in DB.

![uni](https://user-images.githubusercontent.com/748084/99193505-ad7b5800-2779-11eb-900c-68c303b548ac.png)

Moreover, it may be useful to forbid already existing names.
Also, I think that we should not allow the use of a comma "," symbol inside a domain/tag/author name. Because the comma is used to delimiter two different names when someone upload or edit a paper.

Devhub01 replied:
I fixed this by deleting the autocomplete function, and testing the existence of the domain using JavaScript. 
This however made it impossible to modify the domain name if it was written for example in capital letter or vice-versa (kdd=KDD). 
Perhaps, we can add a JS function to transform the text value into for example first capital letter of each word(complex networks to  Complex Networks)  when adding a paper !

kerzol replied:
A good idea, but I cannot say that it will be 100% ok to silently capitalize all first letters.
Maybe, when we test for the existence we should all domains except the current ?

Devhub01 replied:
Okay, I'll work on it

kerzol commented on papersite/catalog.py:
```diff
@@ -1,132 +1,365 @@
+        dom = query_db("select * from domains where domainname = ?",
+                       [domainname], one=True)
+        delete_domain(domainname)
+        delete_papers_domain(dom['domainid'])
```
It looks a little bit strange to delete a domain (or a tag, or an author's name) when there exist papers that use this domain. Sure, it is powerful, but potentially dangerous operation.
An admin can accidentally remove the domain together with all paper-domain relations !
It seems to me that we should not allow the deletion of a domain (or a tag, or an author's name) if there exist some papers linked to it. Or at least display a warning, before proceeding 

 What do you think ?

kerzol reviewed (changes requested):
It works and works well ! 

Below I give my comments related to the user interface style, code style and some comments on the functional part.

Good gob!

Commit:
Updated the last task concerning the domains, authors and tags

Devhub01 committed:
Update edit.html

Devhub01 committed:
Update catalog.py

updated the queries sent in edit template to check the domain/tag/author except the current one

kerzol commented:
ok, If you are ready. I propose to merge it.

Devhub01 commented:
Yes please.

kerzol commented:
" Only those with write access to this repository can merge pull requests. " Could you please merge it or give me a write access to your repo?

kerzol commented:
To be clear: I propose to merge develop into master inside your repo, than propagate changes into my repo via pull-request.

Devhub01 commented:
Normally I have added you with full control to the repo last time.
I have sent you a new invitation so you can do the pull request, I'm still not quite familiar with git commands

Devhub01 commented:
I think I did it, I merged the two branches inside my repo, and create it a pull request to your repo.
